File: Model/Scientific work/ModelPoisson.cpp
#include "ModelPoisson.h"

#include <cmath>
#include <iostream>

ModelPoisson::ModelPoisson() : time(0), lambda(0) {
}

ModelPoisson::ModelPoisson(double lambda_, double time_) : lambda(lambda_), time(time_) {
}

size_t ModelPoisson::countRequests() {
    size_t count = 0;
    double p = generateProbability();
    std::cout << "Random = " << p;
    p *= exp(lambda * time);
    if (abs(1 - p) <= eps)
        return 0;
    double l = 1;
    double F = l;
    while (F <= p) {
        count++;
        l *= lambda * time / count;
        if (l <= eps)
            return count;
        F += l;
    }
    return count;
}

bool ModelPoisson::isCorrect(int count_requests_, double time_) const {
    return std::abs(count_requests_ / time_ - lambda) < 0.1 * lambda;
}

File: cars_pack_py/backend/model_poisson.py
import numpy
import math

from .model import Model
from . import consts


class ModelPoisson(Model):
    def __init__(self, lamb=0, time=0) -> None:
        super().__init__()
        if not (0 <= lamb <= 1):
            raise ValueError()
        self.lamb = lamb
        self.time = time

    def _is_correct(self, count_requests: int, time: float) -> bool:
        return abs(count_requests / time - self.lamb) < 0.1 * self.lamb

    def count_requests(self) -> int:
        count = 0
        a = self.lamb * self.time
        p = numpy.random.uniform(0, 1 - consts.EPSILON) * math.exp(a)
        l = 1
        F = l
        while F <= p:
            count += 1
            l *= a / count
            if l <= consts.EPSILON:
                return count
            F += l
        return count
